README.md
## Screen shots ##

<table align="center" hspace="10" vspace="10" width="100%">
<tr>
    <td>
    <center><img src="images/agenda.gif" /></a></center>
    </td>
    <td>
    <center><img src="images/day.gif"/></a></center>
    </td>
</tr>
<tr>
    <td>
    <center><img src="images/week.gif"/></a></center>
    </td>
    <td>
    <center><img src="images/monthly.gif"/></a></center>
    </td>
</tr>
<tr>
    <td>
    <center><img src="images/finishitem.gif"/></a></center>
    </td>
    <td>
    <center><img src="images/completion.gif"/></a></center>
    </td>
</tr>
<tr>
    <td>
    <center><img src="images/openfile.gif"/></a></center>
    </td>
    <td>
    <center><img src="images/newfile.gif"/></a></center>
    </td>
</tr>
<tr>
    <td>
    <center><img src="images/date_calculator_entry.gif"/></a></center>
    </td>
    <td>
    <center><img src="images/date_calculator_result.gif"/></a></center>
    </td>
</tr>
<tr>
    <td>
    <center><img src="images/availabledates1.gif"/></a></center>
    </td>
    <td>
    <center><img src="images/availabledates2.gif"/></a></center>
    </td>
</tr>
<tr>
    <td>
    <center><img src="images/custom0.gif"/></a></center>
    </td>
    <td>
    <center><img src="images/custom3.gif"/></a></center
    </td>
</tr>
<tr>
    <td colspan=2>
    <center><img src="images/customCSV.gif"/></a></center>
    </td>
</tr>
</table>

<hr/>

## Sample entries ##

Items in *etm* begin with a type character such as an asterisk (event) and continue on one or more lines either until the end of the file is reached or another line is found that begins with a type character. The beginning type character for each item is followed by the item summary and then, perhaps, by one or more `@key value` pairs. The order in which such pairs are entered does not matter.

* A sales meeting (an event) [s]tarting seven days from today at 9:00am and [e]xtending for one hour with a default [a]lert 5 minutes before the start:

        * sales meeting @s +7 9a @e 1h @a 5

* The sales meeting with another [a]lert 2 days before the meeting to (e)mail a reminder to a list of recipients:

        * sales meeting @s +7 9a @e 1h @a 5
          @a 2d: e; who@when.com, what@where.org

* Prepare a report (a task) for the sales meeting [b]eginning 3 days early:

        - prepare report @s +7 @b 3

* A period [e]xtending 35 minutes (an action) spent working on the report yesterday:

        ~ report preparation @s -1 @e 35

* Get a haircut (a task) on the 24th of the current month and then [r]epeatedly at (d)aily [i]ntervals of (14) days and, [o]n completion,  (r)estart from the completion date:

        - get haircut @s 24 @r d &i 14 @o r

* Payday (an occasion) on the last week day of each month. The `&s -1` part of the entry extracts the last date which is both a weekday and falls within the last three days of the month):

        ^ payday @s 1/1 @r m &w MO, TU, WE, TH, FR &m -1, -2, -3 &s -1

* Take a prescribed medication daily (a reminder) [s]tarting today and [r]epeating (d)aily at [h]ours 10am, 2pm, 6pm and 10pm [u]ntil (12am on) the fourth day from today. Trigger the default [a]lert zero minutes before each reminder:

        * take Rx @s +0 @r d &h 10, 14, 18, 22 &u +4 @a 0

* Move the water sprinkler (a reminder) every thirty mi[n]utes on Sunday afternoons using the default alert zero minutes before each reminder:

        * Move sprinkler @s 1 @r w &w SU &h 14, 15, 16, 17 &n 0, 30 @a 0

    To limit the sprinkler movement reminders to the [M]onths of April through September each year change the @r entry to this:

        @r w &w SU &h 14, 15, 16, 17 &n 0, 30 &M 4, 5, 6, 7, 8, 9

     or this:

        @r n &i 30 &w SU &h 14, 15, 16, 17 &M 4, 5, 6, 7, 8, 9

* Presidential election day (an occasion) every four years on the first Tuesday after a Monday in November:

        ^ Presidential Election Day @s 2012-11-06
          @r y &i 4 &M 11 &m 2, 3, 4, 5, 6, 7, 8 &w TU

* Join the etm discussion group (a task) [s]tarting on the first day of the next month. Because of the @g (goto) link, pressing Ctrl-G when the details of this item are displayed in the gui would open the link using the system default application which, in this case, would be your default browser:

        - join the etm discussion group @s +1/1
          @g http://groups.google.com/group/eventandtaskmanager/topics

## Installation ##

### Source installation under OS X, Linux or Windows

Python 2.7.x or python >= 3.3.0 is required.

The following python packages are required for etm but are not included in the python standard library:

-   dateutil (1.5 is OK but >= 2.1 is strongly recommended)
-   PyYaml (>= 3.10)
-   icalendar (>=3.5 for python 2, >= 3.6 for python 3)

Tk and the python module tkinter are also required but are typically already installed on most modern operating systems. If needed, installation instructions are given at www.tkdocs.com/tutorial/install.html.

### Installing etm

Download 'etmtk-x.x.x.tar.gz' from this site, unpack the tarball, cd to the resulting directory and do the normal

    sudo python setup.py install

for a system installation. You can then run from any directory either

    $ etm ?

for information about command line usage or

    $ etm

to open the etm gui.

Alternatively, you can avoid doing a system installation and simply run either

    $ python etm ?

or

    $ python etm

or

    $ ./etm

from this directory.

### Installing Git or Mercurial

*Having one of these version control systems is optional but strongly recommended!*

With either progam installed, etm will automatically commit any change made to any data file. You can see the history of your changes either to a specific file or to any file from the GUI and, of course, you have the entire range of possibilities for showing changes, restoring previous versions and so forth from the command line.

#### Git

Download Git from

    http://git-scm.com/downloads

Install git and then in a terminal enter your personal information

    $ git config --global user.name "John Doe"
    $ git config --global user.email johndoe@example.com

the editor you would like to use

    $ git config --global core.editor vim

and the diff program

    $ git config --global merge.tool vimdiff

Usage information can be obtained in several ways from the terminal

    $ git help <verb>
    $ git <verb> --help
    $ man git-<verb>

Finally, *Pro Git* by Scott Chacon is available to read or download at:

    http://git-scm.com/book/en

If you have been using Mercurial and would like to give Git a try, you can import your etm Mercurial records into Git as follows:

    $ cd
    $ git clone git://repo.or.cz/fast-export.git
    $ git init new_temp_repo
    $ cd new_temp_repo
    $ ~/fast-export/hg-fast-export.sh -r /path/to/etm/datadir
    $ git checkout HEAD

If an "unnamed head" error is reported, try adding `--force` to the end of the fast-export line.

At this point, you should have a copy of your etm datadir in `new_temp_repo` along with a directory, `.git`, that you can copy to the root of your etm datadir where it will join its Mercurial counterpart, `.hg`. You can then delete `new_temp_repo`.

You can now open `etmtk.cfg` for editing and change the setting for `vcs_system` to

    vcs_system: git

#### Mercurial

Download Mercurial from

    http://mercurial.selenic.com/

install it and then create the file *~/.hgrc*, if it doesn't already exist, with at least the following two lines:

    [ui]
    username = Your Name <your email address>


### New etm users

By default, etm will use the directory

    ~/.etm

The first time you run etm it will create, if necessary, the following:

    ~/.etm/
    ~/.etm/etmtk.cfg
    ~/.etm/etmtk_log.txt
    ~/.etm/data/


If the data directory needs to be created, then the following structure will be added:

    ~/.etm/data/
        personal/
            monthly/
        sample/
            completions.cfg
            reports.cfg
            sample.txt
            users.cfg
        shared/
            holidays.txt


The files `sample.txt` and `holidays.txt` contain illustrative data entries and the `*.cfg` files contain illustrative configuration entries.

The following entry will also be inserted in `etmtk.cfg`:

        calendars:
        - - personal
          - true
          - personal
        - - sample
          - true
          - sample
        - - shared
          - true
          - shared

to illustrate the use of calendars.

### Previous etm users

The first time you run etm, it will copy your current configuration settings from `~/.etm/etm.cfg` to `~/.etm/etmtk.cfg`. You can make any changes you like to the latter file without affecting the former.

You can switch back and forth between etm_qt and etm. Any changes made to your data files by either one will be compatible with the other one.


## Feedback ##

Please share your ideas in the discussion group at [GoogleGroups][].

[GoogleGroups]: http://groups.google.com/group/eventandtaskmanager

## Version numbers ##

*etm*'s version numbering uses the `major.minor.patch` format where each of the three components is an integer:

- Major version numbers change whenever there is a large or potentially backward-incompatible change.

- Minor version numbers change when a new, minor feature or a set of smaller features is introduced or when a status change has occured.

- Patch numbers change for new builds involving small bugfixes or the like. Some new builds may not be released.

When the major version number is incremented, both the minor version number and patch number are reset to zero. Similarly, when the minor version number is incremented, the patch number is reset to zero. All increments are by one.

## License ##

Copyright (c) 2009-2014 [Daniel Graham]. All rights reserved.

[Daniel Graham]: mailto://daniel.graham@duke.edu

This program is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License] as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version.

[GNU General Public License]: http://www.gnu.org/licenses/gpl.html



CHANGES
# Changes in the 2 weeks preceding Fri Jun 13 06:30:32 EDT 2014:
- 2014-06-13 06:31:06 -0400 (HEAD, tag: 3.0.11, master): Dan Graham
   tagged version 3.0.11
- 2014-06-13 06:24:32 -0400 (origin/master): Dan Graham
   Tweaks for mk_docs and upload.
- 2014-06-13 06:12:46 -0400: Dan Graham
   Merge branch 'master' of https://github.com/dagraham/etm-tk
- 2014-06-12 22:28:54 -0400 (tag: 3.0.13): Dan Graham
   tagged version 3.0.13
- 2014-06-12 22:28:41 -0400: Dan Graham
   Version number catchup.
- 2014-06-12 22:13:48 -0400: Dan Graham
   moved help files to etm-tk.wiki.
- 2014-06-12 17:43:23 -0400 (tag: 3.0.12): Dan Graham
   tagged version 3.0.12
- 2014-06-12 16:28:49 -0400 (tag: 3.0.9-7): Dan Graham
   tagged version 3.0.11
- 2014-06-12 16:24:46 -0400: Dan Graham
   Keep UserManual pdf and html.
- 2014-06-12 15:55:01 -0400: Dan Graham
   Updated version numbers and added INDEX.md.
- 2014-06-12 15:52:17 -0400: Dan Graham
   Documentation reorganization.
- 2014-06-12 10:14:39 -0400 (tag: 3.0.10): Dan Graham
   tagged version 3.0.10
- 2014-06-12 10:14:26 -0400: Dan Graham
   Added tkinter >= 8.5.11 to REQUIRES.
- 2014-06-12 10:02:25 -0400: Dan Graham
   Removed Windows XP from the trove list of supported platforms.
- 2014-06-12 08:14:13 -0400: Dan Graham
   Documentation updates.
- 2014-06-11 16:16:39 -0400 (tag: 3.0.9): Dan Graham
   tagged version 3.0.9
- 2014-06-11 16:02:36 -0400: Dan Graham
   Added warning that timezone information is ignored to result from
   date calculator for pre 1970 dates under Windows.
- 2014-06-11 13:00:43 -0400: Dan Graham
   Eliminated use of the .etmtkdata.pkl file and, with it, the need to
   import pickle and the functions dump_data and load_data.  Fixed bug
   in date calculator regex. Restored ability to use "easter(yyyy)" in
   date calculator.
- 2014-06-10 16:55:19 -0400 (tag: 3.0.8): Dan Graham
   tagged version 3.0.8
- 2014-06-10 15:24:52 -0400: Dan Graham
   Added hack to enable Windoz to deal with pre 1970-01-01 datetimes. 
   Corrected argument call to get_changes.
- 2014-06-10 11:44:10 -0400 (tag: 3.0.7): Dan Graham
   tagged version 3.0.7
- 2014-06-10 11:40:02 -0400: Dan Graham
   Only use logfile on startup if etmdir exists.  Dump and load
   uuid2labels in pickle file.  Only remove pickle file with bad data
   if it exists. Added &E, byeaster, to rrule options.
- 2014-06-09 13:03:11 -0400: Dan Graham
   Added Shift-Ctrl-F shortcut for clear filter.
- 2014-06-09 10:12:35 -0400: Dan Graham
   Added missing nullok to opts for startActionTimer.
- 2014-06-09 07:06:24 -0400 (tag: 3.0.6): Dan Graham
   tagged version 3.0.6
- 2014-06-09 07:05:22 -0400: Dan Graham
   Initialize edithsh to avoid error when editing files.
- 2014-06-08 12:58:42 -0400: Dan Graham
   Fixed bug in reference to self.options instead of loop.options in
   edit.py.
- 2014-06-08 07:38:43 -0400: Dan Graham
   Delete content on switch to either week or month view.
- 2014-06-07 07:22:09 -0400 (tag: 3.0.5): Dan Graham
   tagged version 3.0.5
- 2014-06-07 07:21:55 -0400: Dan Graham
   Use isocalendar week numbers in month view (as in week view).
   Updated screenshots.
- 2014-06-07 06:39:47 -0400 (tag: 3.0.4): Dan Graham
   tagged version 3.0.4
- 2014-06-06 17:36:08 -0400: Dan Graham
   Make sure tree has focus on startup, even if there are no data
   items.
- 2014-06-06 15:43:35 -0400: Dan Graham
   Remove confirmation for new event on double-click in week view.
- 2014-06-06 15:28:05 -0400: Dan Graham
   Don't insert ? and @e 1h when creating a new event in week view.
- 2014-06-06 15:17:24 -0400: Dan Graham
   When pressing "N" to create a new item in day, week or month view,
   start with @s set to the currently selected date. Also for Return
   or Double-Click in month view.
- 2014-06-06 11:40:07 -0400 (tag: 3.0.3): Dan Graham
   tagged version 3.0.3
- 2014-06-06 11:39:29 -0400: Dan Graham
   Fixed bug in closeMonthly.
- 2014-06-06 10:53:47 -0400: Dan Graham
   Create etmtk_log.txt if it does not exist in get_options.
- 2014-06-06 10:41:31 -0400: Dan Graham
   Added busy and free times lists to month view.
- 2014-06-05 16:29:15 -0400: Dan Graham
   Updated images and README.
- 2014-06-05 13:06:42 -0400 (tag: 3.0.1): Dan Graham
   tagged version 3.0.1
- 2014-06-05 13:05:46 -0400: Dan Graham
   Added monthly screenshot.
- 2014-06-05 13:05:22 -0400: Dan Graham
   Respect active calendars in month view. Updated shortcuts. Added
   Month view to overview.
- 2014-06-05 12:03:37 -0400: Dan Graham
   Month view works with up, down, left, right and space bindings. Day
   number color reflects number of minutes scheduled.
- 2014-06-03 18:56:54 -0400: Dan Graham
   Background color tweaks. Hide insertion cursor in ReadOnlyText.
- 2014-06-03 06:52:27 -0400: Dan Graham
   Merge branch 'master' of https://github.com/dagraham/etm-tk
- 2014-06-03 06:50:31 -0400 (tag: 3.0.0): Dan Graham
   tagged version 3.0.0
- 2014-06-03 06:49:06 -0400: Dan Graham
   Restart version numbering at 3.0.0.
- 2014-06-03 06:43:05 -0400: Dan Graham
   Changed idle_minum to idle_minutes.  Changed development status to
   5 - Production/Stable.
- 2014-06-02 22:50:06 -0400: Dan Graham
   Merge branch 'master' of https://github.com/dagraham/etm-tk
- 2014-06-02 22:48:09 -0400 (tag: 0.2.1): Dan Graham
   tagged version 0.2.1
- 2014-06-02 22:47:34 -0400: Dan Graham
   Changed development status to 5 - Production/Stable and version to
   0.2.0.
- 2014-06-02 22:31:44 -0400: Dan Graham
   tagged version 0.2.1
- 2014-06-02 22:31:12 -0400: Dan Graham
   Changed development status to 5 - Production/Stable and version to
   0.2.0.
- 2014-06-02 14:59:46 -0400 (tag: 0.1.44): Dan Graham
   tagged version 0.1.44
- 2014-06-02 14:59:35 -0400: Dan Graham
   Fixed bug in passing font to DialogWindow.
- 2014-06-02 13:15:10 -0400 (tag: 0.1.43): Dan Graham
   tagged version 0.1.43
- 2014-06-02 13:14:46 -0400: Dan Graham
   Added available dates screenshots.
- 2014-06-02 13:13:53 -0400: Dan Graham

main.py
# -*- coding: utf-8 -*-
import sys
import etmKv.data as data
from etmKv.data import get_current_time, init_localization, str2hsh, hsh2str

from kivy.app import App
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.textinput import TextInput
from kivy.uix.codeinput import CodeInput
from kivy.uix.button import Button
# from kivy.uix.widget import Widget

from pygments.lexers.special import TextLexer

from kivy.uix.modalview import ModalView
from kivy.properties import ObjectProperty
from kivy.config import Config
from kivy.clock import Clock
Config.set('graphics', 'height', '420')
Config.set('graphics', 'width', '520')

import gettext
_ = gettext.gettext


class ETMEditor(CodeInput):

    def __init__(self, parent=None, callback=None):
        self.parent = parent
        self.button = Button
        self.callback = callback
        super(ETMEditor, self).__init__(multiline=True, size_hint=(1, None), height=300, lexer=TextLexer())
        self.modified = False
        self.bind(focus=self.on_focus)

    def _keyboard_on_key_up(self, window, keycode):
        # inform ETMDialog about the modification state
        if self.callback:
            self.callback((len(self._undo) > 0))
        super(ETMEditor, self)._keyboard_on_key_up(
            window, keycode)

    def on_focus(self, instance, value):
        if value:
            print('Editor focused', instance)
        else:
            print('Editor defocused', instance)
        super(ETMEditor, self).on_focus(instance, value)


class ETMDialog():

    def __init__(self, parent=None, options={}):
        self.parent = parent
        self.options = options
        self.editor = ETMEditor()
        self.text = ''

        self.btnchecksave = Button(text=_('Close'), size_hint_y=None, height='30sp')
        self.btncancel = Button(text=_('Cancel'), size_hint_y=None, height='30sp')
        buttons = BoxLayout(orientation='horizontal', height='30sp')
        buttons.add_widget(self.btncancel)
        buttons.add_widget(self.btnchecksave)
        self.message = TextInput(text="", size_hint_y=None, height=80)
        self.message.readonly = True
        self.message.background = [1, 1, 1, 1]
        self.content = BoxLayout(orientation='vertical')
        self.content.padding = 3
        self.content.spacing = 3

        self.input = ETMEditor(callback=self.set_modified)
        self.input.background_color = [1, 1, 1, 1]
        self.input.font_size = self.options['fontsize']

        self.content.add_widget(self.input)
        self.content.add_widget(self.message)
        self.content.add_widget(buttons)
        self.popup = ModalView(size_hint=(None, None), size=(500, 420))
        self.popup.add_widget(self.content)
        self.btnchecksave.bind(on_release=self.check_save)
        self.btncancel.bind(on_release=self.cancel)

    def run(self, text="", msg=""):
        print('run', text, msg)
        self.input.text = text
        self.message.text = msg
        self.input.reset_undo()
        self.input.focus = True
        print('opening popup')
        self.popup.open()

    def cancel(self, value):
        self.parent.mode = 'command'
        self.input.text = ''
        self.popup.dismiss()

    def check_save(self, value):
        new_str = ''
        if self.modified:
            print('check_save', self.input.text)
            new_hsh, msg = str2hsh(self.input.text, options=self.options)
            if msg:
                self.message.text = "\n".join(msg)
            else:
                try:
                    new_str = hsh2str(new_hsh, self.options)
                    # print('new_str', new_str, self.parent.mode)
                    new_item = unicode(u"{0}".format(hsh2str(new_hsh, self.options)))
                    self.input.text = new_item
                    if self.parent.mode == 'append':
                        self.parent.loop.append_item(new_hsh, new_item)
                    elif self.parent.mode == 'replace':
                        self.parent.loop.replace_item(new_hsh, new_item)
                    # self.message.text = self.parent.loop.do_n(arg_str=new_item)
                    self.set_modified(False)
                except:
                    self.message.text = "Error processing %s" % new_hsh
        else:  # close
            if new_str:
                self.parent.output_wid.text = new_str
            self.cancel(value)

    def set_modified(self, bool):
        # print('set_modified', bool)
        self.modified = bool
        if bool:
            self.btnchecksave.text = _('Save')
        else:
            self.btnchecksave.text = _('Close')


class ETMTextInput(TextInput):

    history = []
    index = 0
    now = None
    options = {}
    popup = ''
    value = ''
    firsttime = True
    mode = 'command'   # or edit or delete
    item_hsh = {}

    def _keyboard_on_key_down(self, window, keycode, text, modifiers):
        # print(keycode)
        if keycode[0] == 273:  # 273 is the keycode for up
            self.previous_history()
        elif keycode[0] == 274:  # 274 is the keycode for down
            self.next_history()
        elif keycode[0] == 13:  # 13 is the keycode for enter
            self.process_input()
        elif keycode[0] == 27:  # escape
            self.select_all()
            self.delete_selection()
            # return True to avoid closing app
            return(True)
        else:
            super(ETMTextInput, self)._keyboard_on_key_down(
                window, keycode, text, modifiers)

    def init(self):
        if self.firsttime:
            # only do this once
            print('init one time only')
            self.firsttime = False
            self.bind(focus=self.on_focus)

            self.loop = loop
            self.options = self.loop.options
            self.Dialog = ETMDialog(parent=self, options=self.options)
            self.loop.parent = self
            res = self.loop.do_command('a')
            self.start_timer()
            return(res)
        return()

    def start_timer(self):
        self.now = get_current_time()
        nxt = (60 - self.now.second)
        print(self.now)
        Clock.schedule_once(self.timer_callback, nxt)

    def timer_callback(self, dt):
        self.start_timer()

    def get_input(self, question):
        self.command_mode = False
        return(question)

    def process_input(self):
        """
        This is called whenever enter is pressed in the input field.
        Action depends upon comand_mode.
        Append input to history, process it and show the result in output.
        """
        cmd = self.text.strip()

        # if not cmd:
        #     return(True)

        if self.mode == 'command':
            if cmd == 'q':
                sys.exit()
            cmd = cmd.strip()
            if cmd[0] in ['a', 'r', 't']:
                # simple command history for report commands
                if cmd in self.history:
                    self.history.remove(cmd)
                self.history.append(cmd)
                self.index = len(self.history) - 1
            # select everything in input to make it easy to clear
            self.select_all()
            try:
                res = loop.do_command(cmd)
            except:
                return _('could not process command "{0}"').format(cmd)

        elif self.mode == 'edit':
            print('edit', cmd)
            res = loop._do_edit(cmd)

        elif self.mode == 'delete':
            print('deleted', cmd)
            loop._do_delete(cmd)
            res = ''

        elif self.mode == 'finish':
            print('finish', cmd)
            loop._do_finish(cmd)
            res = ''

        elif self.mode == 'new_date':
            print('date', cmd)
            res = loop._new_date(cmd)

        if not res:
            return(True)
        print('display and scroll to top')
        self.output_wid.text = res
        self.scroll_wid.scroll_y = 1
        self.scroll_to_top()

    def scroll_to_top(self):
        # print('scrolling')
        self.scroll_wid.update_from_scroll()

    def previous_history(self):
        """
        Replace input with the previous history item.
        """
        if self.index >= 1:
            self.index -= 1
            self.text = self.history[self.index]

    def next_history(self):
        """
        Replace input with the next history item.
        """
        if self.index + 1 < len(self.history):
            self.index += 1
            self.text = self.history[self.index]

    def on_focus(self, instance, value):
        if value:
            print('focused', instance)
        else:
            print('defocused', instance)
        super(ETMTextInput, self).on_focus(instance, value)


class BoxIOWidget(BoxLayout):
    input_wid = ObjectProperty()
    status_wid = ObjectProperty()
    output_wid = ObjectProperty()
    scroll_wid = ObjectProperty()


class etmApp(App):
    title = 'etm'
    icon = 'etmlogo_512x512x32.png'

    def build(self):
        return BoxIOWidget()

if __name__ == "__main__":
    init_localization()
    etmdir = ''
    if len(sys.argv) > 1:
        etmdir = sys.argv.pop(1)
    (user_options, options, use_locale) = data.get_options(etmdir)
    loop = data.ETMCmd(options)
    etmApp().run()
